File: Cpp/720_Longest_Word_in_Dictionary.cpp
class Trie {
public:
    Trie() : root(new TrieNode()){ 
    }
    
    void insert(const string& word) {
        TrieNode* node = root;
        for(auto& ch : word){
            int index = ch -'a';
            if(node->children[index] == nullptr){
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->is_leaf = true;
    }
    
    bool search(const string& word) {
        TrieNode* node = root;
        for(auto& ch : word){
            int index = ch-'a';
            if(node->children[index] == nullptr) return false;
            node = node->children[index];
        }
        if(word.empty()) return true;
        return node->is_leaf;
    }
    
private:
    struct TrieNode {
        bool is_leaf;
        vector<TrieNode*> children;

        TrieNode() : is_leaf(false), children(26, nullptr){
        }
    };
    
    TrieNode* root;
};

class Solution {
public:
    string longestWord(vector<string>& words) {
        string result = "";
        Trie trie;
        sort(words.begin(), words.end());

        for(auto& word : words){
            if(trie.search(word.substr(0, word.size()-1))){
                if(result.size() < word.size()){
                    result = word;
                }
                trie.insert(word);
            }
            
        }
        return result;
    }
};

File: Kotlin/720_Longest_Word_in_Dictionary.kt
data class Node(
    val children: Array<Node?> = arrayOfNulls(26),
    var isWord: Boolean = false)

class Trie {
    val root = Node()
    
    fun add(word:String) {
        var current = root
        for(ch in word) {
            if(current.children[ch - 'a'] == null) current.children[ch - 'a'] = Node()
            current = current.children[ch - 'a']!!
        }
        current.isWord = true
    }

    fun search(word: String): Boolean{
        var node = root
        for(ch in word) {
            if(node.children[ch - 'a'] == null) return false
            node = node.children[ch - 'a']!!
        }
        if(word.isEmpty()) return true
        return node.isWord
    }
}

class Solution {
    fun longestWord(words: Array<String>): String {
        var result = ""
        val trie = Trie()
        words.sort()
        
        for(word in words){
            if(trie.search(word.substring(0, word.length-1))){
                if(result.length < word.length){
                    result = word
                }
                trie.add(word)
            }
            
        }
        
        return result
    }
}